page-tracking/src/page_tracker.rs

```diff
@@ -516,79 +708,144 @@ impl PageTracker {
+                    && ((info.owner() == Some(owner) && info.state() == PageState::Mapped)
+                        || (info.owner() != Some(owner) && info.is_shared())))
+                {
+                    return false;
+                }
+            } else {
+                return false;
+            }
         }
+        true
     }
 
     /// Returns true if the page is considered removable.
     pub fn is_removable_page(
         &self,
         addr: SupervisorPageAddr,
+        page_size: PageSize,
         owner: PageOwnerId,
         mem_type: MemType,
         tlb_version: TlbVersion,
     ) -> bool {
         let mut page_tracker = self.inner.lock();
-        if let Ok(info) = page_tracker.get(addr) {
-            info.is_removable(tlb_version)
-                && info.mem_type() == mem_type
-                && ((info.owner() == Some(owner) && info.is_blocked())
-                    || (info.owner() == Some(PageOwnerId::host()) && info.is_blocked_shared()))
-        } else {
-            false
+        for paddr in addr
+            .iter_from()
+            .take(PageSize::num_4k_pages(page_size as u64) as usize)
+        {
+            if let Ok(info) = page_tracker.get(paddr) {
+                if !(info.is_removable(tlb_version)
+                    && info.mem_type() == mem_type
+                    && ((info.owner() == Some(owner) && info.is_blocked())
+                        || (info.owner() == Some(PageOwnerId::host()) && info.is_blocked_shared())))
+                {
+                    return false;
+                }
+            } else {
+                return false;
+            }
         }
+        true
     }
 
     /// Creates a link from page `a` to `b` if neither is already linked.
-    pub(crate) fn link_pages(&self, a: SupervisorPageAddr, b: SupervisorPageAddr) -> Result<()> {
+    pub(crate) fn link_pages(
+        &self,
+        a: SupervisorPageAddr,
+        b: SupervisorPageAddr,
+        page_size: PageSize,
+    ) -> Result<()> {
+        let num_pages = PageSize::num_4k_pages(page_size as u64) as usize;
         let mut page_tracker = self.inner.lock();
-        // We don't need to touch the destination, but we need to make sure it isn't linked.
-        let dst = page_tracker.get(b)?;
-        if dst.next().is_some() {
-            return Err(Error::PageAlreadyLinked);
+        let mut prev = a;
+        for next in a
+            .iter_from()
+            .take(num_pages)
+            .chain(b.iter_from().take(num_pages))
+        {
+            if next == prev {
+                continue;
+            }
+            // We don't need to touch the destination, but we need to make sure it isn't linked.
+            let dst = page_tracker.get(next)?;
+            if dst.next().is_some() {
+                return Err(Error::PageAlreadyLinked);
+            }
+            let src = page_tracker.get_mut(prev)?;
+            src.link(next)?;
+            prev = next;
```

Comment:
so is the idea that pages in the list are always linked at 4k and caller has to decide what granularity they are unlinked at? in that case, the size should probably be part of `PageList`. also could use some documentation.

Reply:
yeah it does. we should try to optimize this by only linking the "head" page together, but that can be deferred til later.